ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3CompleteMultipartUploadHandler.java

```diff
@@ -0,0 +1,366 @@
+      }
+      return completeMultipartUpload;
+    }
+
+    /**
+     * Parse {@link BlobInfo} from the request arguments. This method will also ensure that the correct account and
+     * container objects are attached to the request.
+     * @return the {@link BlobInfo} parsed from the request arguments.
+     * @throws RestServiceException if there is an error while parsing the {@link BlobInfo} arguments.
+     */
+    private BlobInfo getBlobInfoFromRequest() throws RestServiceException {
+      long propsBuildStartTime = System.currentTimeMillis();
+      accountAndContainerInjector.injectAccountContainerForNamedBlob(restRequest, frontendMetrics.putBlobMetricsGroup);
+      BlobProperties blobProperties = RestUtils.buildBlobProperties(restRequest.getArgs());
+      frontendMetrics.blobPropsBuildForNameBlobPutTimeInMs.update(System.currentTimeMillis() - propsBuildStartTime);
+      LOGGER.trace("Blob properties of blob being PUT - {}", blobProperties);
+      return new BlobInfo(blobProperties, ByteBuffer.allocate(0).array());
+    }
+
+    /**
+     * Create a {@link BlobProperties} for the router upload (putBlob or stitchBlob) with a finite TTL such that
+     * orphaned blobs will not be created if the write to the named blob metadata DB fails.
+     * @param blobInfoFromRequest the {@link BlobInfo} parsed from the request.
+     * @return a {@link BlobProperties} for a TTL-ed initial router call.
+     */
+    BlobProperties getPropertiesForRouterUpload(BlobInfo blobInfoFromRequest) {
+      BlobProperties properties;
+      if (blobInfoFromRequest.getBlobProperties().getTimeToLiveInSeconds() == Utils.Infinite_Time) {
+        properties = new BlobProperties(blobInfoFromRequest.getBlobProperties());
+        // For blob with infinite time, the procedure is putBlob with a TTL, record insert to database with
+        // infinite TTL, and ttlUpdate.
+        properties.setTimeToLiveInSeconds(frontendConfig.permanentNamedBlobInitialPutTtl);
+      } else {
+        properties = blobInfoFromRequest.getBlobProperties();
+      }
+      return properties;
+    }
+
+    /**
+     * @param throwable the error to check.
+     * @return true if the router error is retriable.
+     */
+    private boolean isRetriable(Throwable throwable) {
+      return throwable instanceof RouterException && retriableRouterError.contains(
+          ((RouterException) throwable).getErrorCode());
+    }
+
+    /**
+     * Parse and verify the part IDs in the body of a CompleteMultipartUpload request.
+     * @param completeMultipartUpload {@link CompleteMultipartUpload} request
+     * @return a list of chunks to stitch that can be provided to the router.
+     */
+    List<ChunkInfo> getChunksToStitch(CompleteMultipartUpload completeMultipartUpload) {
+      List<String> partIds = new ArrayList<>();
+      for (Part part : completeMultipartUpload.getPart()) {
+        partIds.add(part.geteTag());
```

Comment:
Sorry, yeah, this part is actually a TODO since we need to deserialize the ETag and get nested Chunks. I think the steps would be:

1. Rely on partNumber to get ordered list of ETags
2. Deserialize each ETag to get nested chunks
3. Pass the ordered nested chunks to PutOperation

In Ambry, after we write all the chunks to server, we write the metadata file containing the chunk IDs. In the metadata file, I think we make sure to capture the ordering of chunk IDs - [code](https://github.com/linkedin/ambry/blob/6a64918649ae95ddc48578623dd443b00a01e949/ambry-router/src/main/java/com/github/ambry/router/PutOperation.java#L418)

Reply:
Thanks for the clarification!